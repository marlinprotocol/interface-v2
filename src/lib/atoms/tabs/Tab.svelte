<script lang="ts">
	import { staticImages } from '$lib/components/images/staticImages';
	import { cn } from '$lib/utils/helpers/commonHelper';

	export let id: string;
	export let liClass = 'mr-2';
</script>

<li
	class={cn('relative flex w-full list-none justify-center', liClass)}
	role="presentation"
	data-testid="tab"
>
	<button
		on:click
		class="-my-5 flex h-14 w-14 items-center justify-center rounded-full border border-grey-100 bg-white"
		id="{id}-tabhead"
		type="button"
		role="tab"
	>
		<img src={staticImages.arrowConversion} alt="Conversion" />
	</button>
</li>
