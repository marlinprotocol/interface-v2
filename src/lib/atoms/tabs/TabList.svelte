<script>
	export let divClass = '';
</script>

<div class={divClass} data-testid="tab-list">
	<ul class="-mb-px flex flex-wrap items-center justify-center gap-1" role="tablist">
		<slot />
	</ul>
</div>
