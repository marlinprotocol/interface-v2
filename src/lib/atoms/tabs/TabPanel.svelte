<script lang="ts">
	export let activeTabValue: string;
	export let id: string;
</script>

{#if activeTabValue === id}
	<div
		class="w-full"
		id="{id}-tabitem"
		role="tabpanel"
		aria-labelledby="{id}-tab"
		data-testid="tab-panel"
	>
		<slot />
	</div>
{/if}
