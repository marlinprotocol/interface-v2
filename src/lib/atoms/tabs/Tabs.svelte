<script lang="ts">
	export let divClass = 'w-full';
</script>

<div class={divClass} data-testid="tabs">
	<slot />
</div>
