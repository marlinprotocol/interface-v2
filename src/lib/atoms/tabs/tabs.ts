export { default as Tabs } from '$lib/atoms/tabs/Tabs.svelte';
export { default as TabList } from '$lib/atoms/tabs/TabList.svelte';
export { default as TabPanel } from '$lib/atoms/tabs/TabPanel.svelte';
export { default as Tab } from '$lib/atoms/tabs/Tab.svelte';
