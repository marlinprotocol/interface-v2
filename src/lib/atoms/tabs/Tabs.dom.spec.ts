import { cleanup, render } from '@testing-library/svelte';
import html from '@playpilot/svelte-htm';
import Tab from './Tab.svelte';
import { TabList, TabPanel, Tabs } from './tabs';

describe('Tabs', () => {
	afterEach(() => {
		cleanup();
		vi.resetAllMocks();
	});

	it('renders the tab', () => {
		const { getByTestId } = render(Tab, { props: { id: 'tab_test' } });
		const tabElement = getByTestId('tab');
		expect(tabElement.childNodes).toHaveLength(1);
		expect((tabElement.firstChild as Element)?.id === 'tab_test-tabhead').toBe(true);
		expect(tabElement).toMatchSnapshot();
	});

	it('renders the tab with slot', () => {
		const { getByTestId } = render(html`<${Tab} id="tab_test">Tab</${Tab}>`);
		const tabElement = getByTestId('tab');
		expect(tabElement).toBeTruthy();
		expect(tabElement.childNodes).toHaveLength(1);
		expect((tabElement.firstChild as Element)?.id === 'tab_test-tabhead').toBe(true);
	});

	it('renders the tab and verifies it using getByTestId & getByRole ', () => {
		const { getByTestId, getByRole } = render(Tab, { props: { id: 'tab_test' } });
		expect(getByTestId('tab')).toEqual(getByRole('presentation'));
	});

	it('renders the tab with all atributes', () => {
		const { getByTestId } = render(Tab, {
			props: { liClass: 'list', id: 'tab_test' }
		});
		const tabElement = getByTestId('tab');
		expect(tabElement.classList.contains('list')).toBeTruthy();
		expect((tabElement.firstChild as Element).classList.contains('bg-white')).toBe(true);
	});

	it('renders the default tabs', () => {
		const { getByTestId } = render(Tabs);
		const tabElement = getByTestId('tabs');
		expect(tabElement.classList.contains('w-full')).toBe(true);
	});

	it('renders the tabs with slot', () => {
		const { getByTestId, getByText } = render(
			html`<${Tabs}><div id="tabs-slot">Tabs</div></${Tabs}>`
		);
		const tabElement = getByTestId('tabs');
		expect(tabElement.classList.contains('w-full')).toBe(true);
		expect(getByText('Tabs').id).toEqual('tabs-slot');
	});

	it('renders the tabs with divClass', () => {
		const { getByTestId } = render(Tabs, { props: { divClass: 'tabs-primary' } });
		const tabElement = getByTestId('tabs');
		expect(tabElement.classList.contains('tabs-primary')).toBe(true);
	});

	it('renders the default tablist', () => {
		const { getByTestId, getByRole } = render(TabList);
		const tabListElement = getByTestId('tab-list');
		expect(tabListElement.contains(getByRole('tablist'))).toBe(true);
		expect(tabListElement).toMatchSnapshot();
	});

	it('renders tablist with divClass', () => {
		const { getByTestId } = render(TabList, { props: { divClass: 'tablist-divclass' } });
		const tabListElement = getByTestId('tab-list');
		expect(tabListElement.classList.contains('tablist-divclass')).toBe(true);
	});

	it('renders tablist with slot', () => {
		const { getByTestId, getByText } = render(
			html`<${TabList} divClass='tablist-divclass' ><div id="tablist-slot">Tablist</div></${TabList}>`
		);
		const tabListElement = getByTestId('tab-list');
		expect(tabListElement.classList.contains('tablist-divclass')).toBe(true);
		expect(getByText('Tablist').id === 'tablist-slot').toBe(true);
	});

	it('does not render TabPanel', () => {
		const { queryAllByTestId } = render(TabPanel, { props: { activeTabValue: '10', id: '20' } });
		expect(queryAllByTestId('tab-panel')).toHaveLength(0);
	});

	it('renders TabPanel', () => {
		const { getByTestId } = render(TabPanel, { props: { activeTabValue: '10', id: '10' } });
		expect(getByTestId('tab-panel').id === '10-tabitem').toBeTruthy();
		expect(getByTestId('tab-panel')).toMatchSnapshot();
	});

	it('renders TabPanel with slot', () => {
		const { getByTestId, getByText } = render(
			html`<${TabPanel} id="10" activeTabValue='10'><div id="tab-panel-slot">TabPanel</div></${TabPanel}>`
		);
		expect(getByTestId('tab-panel').id === '10-tabitem').toBeTruthy();
		expect(getByText('TabPanel').id === 'tab-panel-slot').toBe(true);
	});

	it('renders TabPanel using getByRole', () => {
		const { getByRole } = render(TabPanel, { props: { activeTabValue: '50', id: '50' } });
		expect(getByRole('tabpanel').getAttribute('data-testid') === 'tab-panel').toBe(true);
	});
});
