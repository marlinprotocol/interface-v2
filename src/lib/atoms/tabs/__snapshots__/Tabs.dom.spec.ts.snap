// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tabs > renders TabPanel 1`] = `
<div
  aria-labelledby="10-tab"
  data-testid="tab-panel"
  id="10-tabitem"
  role="tabpanel"
/>
`;

exports[`Tabs > renders the default tablist 1`] = `
<div
  class=""
  data-testid="tab-list"
>
  <ul
    class="-mb-px flex flex-wrap items-center justify-center gap-1"
    role="tablist"
  />
</div>
`;

exports[`Tabs > renders the tab 1`] = `
<li
  class="mr-2"
  data-testid="tab"
  role="presentation"
>
  <button
    class=""
    id="tab_test-tabhead"
    role="tab"
    type="button"
  />
</li>
`;
